### ISSUE TITLE
Gyroscope/compass Linux/Jetson Nano driver

### ISSUE THREAD

**Issue by ntwong0** [SJSURoboticsTeam/urc-central-2021#48]
progress
- [x] get raw data from MPU-9250
- [x] Decode compass data
- [ ] read data with ros-mpu9250-node
- [ ] Determine compass calibration procedure

useful link
- https://github.com/wolfeidau/ros-mpu9250-node

ref for i2c timing diagram
![image](https://user-images.githubusercontent.com/10604448/85188319-908a3980-b25a-11ea-8b46-848b7facb5ae.png)

**Comment by ntwong0** [comment]
No updates for this week, will check back on @BobMak next week

**Comment by BobMak** [comment]
![img](https://i.imgur.com/FxJzZGm.png)
errors in making upm from the source. Need to resolve dependencies. so far tried apt-get install libbacnet libmodbus libopenzwave libopenzwave

**Comment by ntwong0** [comment]
@BobMak can you attempt building `upm` from source by next Saturday? Stretch goal - draft of calibration procedure

**Comment by ntwong0** [comment]
@BobMak @YMmirsky need clarification on next steps

**Comment by ntwong0** [comment]
@BobMak still blocked on `upm` building, will meet with @YMmirsky to resolve

**Comment by BobMak** [comment]
Able to build upm. upm requires a specific nodejs version to build optional utilities. It is possible to build upm without node by running `cmake <upm_directory> DBUILDSWIGNODE=OFF` 
[Original solution](https://github.com/eclipse/upm/issues/225)

**Comment by PushpalPatil** [comment]
@BobMak Note: saturday

**Comment by BobMak** [comment]
Rebuilding upm with raw=TRUE results in the same error

**Comment by BobMak** [comment]
[Core issue](https://github.com/eclipse/mraa/issues/997). Replacing mraa i2c read/write calls in upm.

**Comment by BobMak** [comment]
I've replaced the mraa i2c calls with linux i2c code in the upm as a temporary workaround. I can see the accelerometer data, but the orientation data looks broken.

**Comment by PushpalPatil** [comment]
maybe use magnet method

**Comment by BobMak** [comment]
Tried enabling magnetometer by setting the `enableAk8975` flag to true. Still getting garbage data, no noticeable changes when the magnet is put closer.

**Comment by PushpalPatil** [comment]
can you post an example or vid recording of the garbage data? also could you push your code to the repo so that we can see the code?

**Comment by BobMak** [comment]
[garbage data and magnet test](https://youtu.be/cJh2a9UlK9U)

**Comment by BobMak** [comment]
Failed to utilize the ros-mpu9250-node module on Jetson Nano. The low-level dependency of the node doesn't support Jetson nano yet. Replacing the dependency failed. Working on my own magnetometer node implementation. Already able to initialize and read from the magnetometer.

**Comment by BobMak** [comment]
working on a python mpu9250 orientation node implementation.

**Comment by BobMak** [comment]
mpu orientation data looks consistent and is reacting to a magnet.
now trying to figure out how to do self-test and how to calibrate the sensor.

vizualised pose data by using [Nelson's rviz notes](https://github.com/SJSURoboticsTeam/urc-autonomy-subsystem-2021/tree/mpu9250_viz)
there was one other thing I needed to run to see the live data visualisation:
```rosrun tf static_transform_publisher 0.0 0.0 0.0 0.0 0.0 0.0 map imu 100```

**Comment by BobMak** [comment]
[mpu9250 compass heading visualization](https://www.youtube.com/watch?v=zxfoSmDnpBw)
